src/components/DuplicateCourseModal.js

```diff
@@ -0,0 +1,162 @@
+    ['ui-courses', 'terms'],
+    () => ky(manifest.terms.path).json(),
+    {
+      placeholderData: [],
+      onSuccess: json => setTerm(json[0]?.value),
+      select: json => ((json?.terms ?? [])
+        .map(t => ({ value: t.id, label: t.name }))
+        .sort((a, b) => a.label.localeCompare(b.label))
+      )
+    }
+  );
+
+  const [duplicateCrosslistedCourses, setDuplicateCrosslistedCourses] = useState(false);
+  useQuery(
+    ['ui-courses', 'duplicateCrosslistedCourses'],
+    () => ky(manifest.displaySettings.path).json(),
+    {
+      onSuccess: json => setDuplicateCrosslistedCourses(json.duplicateCrosslistedCourses),
+      select: json => JSON.parse(json.configs[0].value),
+    },
+  );
+
+  const duplicateCourse = async () => {
+    const { course, crossListed, reserves } = data;
+    const { courseListingObject: courseListing } = course;
+
+    const newCourseListing = await ky.post('coursereserves/courselistings', { json: {
+      courseTypeId: courseListing.courseTypeId,
+      externalId: courseListing.externalId,
+      locationId: courseListing.locationId,
+      registrarId: courseListing.registrarId,
+      servicepointId: courseListing.servicepointId,
+      termId: term,
+    } }).json();
+
+    for (let i = 0; i < courseListing.instructorObjects.length; i++) {
+      await ky.post(`coursereserves/courselistings/${newCourseListing.id}/instructors`, { json: {
+        courseListingId: newCourseListing.id,
+        name: courseListing.instructorObjects[i].name,
+        userId: courseListing.instructorObjects[i].userId,
+      } }).json();
+    }
+
+    for (let i = 0; i < reserves.length; i++) {
+      await ky.post(`coursereserves/courselistings/${newCourseListing.id}/reserves`, { json: {
+        courseListingId: newCourseListing.id,
+        itemId: reserves[i].itemId,
+      } }).json();
+    }
+
+    const newCourse = await ky.post(`coursereserves/courselistings/${newCourseListing.id}/courses`, { json: {
+      courseListingId: newCourseListing.id,
+      courseNumber: course.courseNumber,
+      departmentId: course.departmentId,
+      description: course.description,
+      name: `${course.name} - ${intl.formatMessage({ id: 'ui-courses.button.duplicate' })}`,
+      sectionName: course.sectionName,
+    } }).json();
+
+    if (duplicateCrosslistedCourses) {
+      for (let i = 0; i < crossListed.length; i++) {
+        await ky.post(`coursereserves/courselistings/${newCourseListing.id}/courses`, { json: {
+          courseListingId: newCourseListing.id,
+          courseNumber: crossListed[i].courseNumber,
+          departmentId: crossListed[i].departmentId,
+          description: crossListed[i].description,
+          name: `${crossListed[i].name} - ${intl.formatMessage({ id: 'ui-courses.button.duplicate' })}`,
+          sectionName: crossListed[i].sectionName,
+        } }).json();
+      }
+    }
+
+    history.push(`/cr/courses/${newCourse.id}`);
+
+    onClose();
+  };
```

Comment:
Despite quibbles, I really like how this function reads. It's like an advertisement for `async`/`await`. I don't dare contemplate how it would look with explicit promises.